**Joshua Kim**
 
**SpringBoard - Data Science Career Track**

# Capstone Project 2: Topic-Modeling with /r/PersonalFinance

***

## Table of Contents

0. [Introduction](#0.-Introduction)


1. [Data Acquisition](#1.-Data-Acquisition)


2. [Data Wrangling](#2.-Data-Wrangling)


3. [Exploratory Data Analysis](#3.-Exploratory-Data-Analysis)


4. [Machine Learning](#4.-Machine-Learning)

    a) [Part 1: Creating the Topics](#Part-1:-Creating-the-Topics)
    
    b) [Part 2: Evaluating/Predicting the Topics](#Part-2:-Evaluating/Predicting-the-Topics)


5. [Conclusion](#5.-Conclusion)


6. [Resources](#6.-Resources)

***

## 0. Introduction

[Return to Table of Contents](#Table-of-Contents)

>### The Problem

One of the common problems that the everyday American faces is their financial situation. There are many different reasons for this, such as debt, student loans, the passing of a parent. According to a study of 1000 Americans over the age of 30 (conducted by [GuideVine](https://www.guidevine.com/newsroom/half-americans-age-30-cant-explain-401k/), less than half of Americans understood common financial concepts such as interest, bankruptcy, and inflation. More importantly, more than half felt "lost" when it came to having a secure long-term plan for their financial future. This highlights a re-occuring theme in our society where many people have trouble getting by with their finances and naturally some people seek help online. One of these resources is a subreddit called 'Personal Finance', which is hosted on the popular discussion-board website **[Reddit](https://www.reddit.com/)**.

![s](https://github.com/nysportsfan/Personal_Finance_Subreddit/blob/master/Images/PersonalFinance%20Subreddit%20Screenshot.png?raw=true)

**[/r/personalfinance](https://www.reddit.com/r/personalfinance/)** has become a place where many people are able to ask for help regarding their financial situations. A few common topics are savings, budgeting, taxes and debt. There are hundreds of different posts everyday from people of different ages, genders, locations and occupations. Frequent users of the subreddit often provide advice through comments and are able to upvote posts to help it gain more visibility on the front page of /r/personalfinance. However, the main feature we will be focusing on in this project is the "flair" feature which enables users to tag their submissions with a topic. For example, if I wrote a post asking advice on paying off my student loans, I would use the 'Debt' flair. This helps organize the hundreds of daily posts into different categories so that people will be able to search them more easily.

There are 3 main problems with this:

<div class="alert alert-block alert-danger">
1) <b>Not all users choose a flair for their post, which results in a default 'Other' tag on their post.</b> 
</div>

In this scenario, the system will automatically tag the post with 'Other', which is used when the user either doesn't know what topic the post should fall under OR if the user does not assign a topic. This becomes a problem since this can result in many posts not having their proper flair and ruins the organized structure of searching by topic. The 'Other' flair was originally designed to contain posts that did not fall under the other topics but due to the popularity of /r/PersonalFinance (over 13 million subscribers), there are often many new users who do not how to either assign a topic (issue with the interface) or choose a topic (unfamiliar with financial categories). Because of this, the 'Other' flair is usually the most popular and common topic on the subreddit despite the fact that most of the posts actually do fall under another topic. 

It would be like someone throwing their plastic trash into a regular garbage bin instead of the recycle bin because they were unaware of what the recycle bin is for, they did not see it, or they did not care to throw it into the proper bin. If this was simply a rarity, it could be overlooked but if it becomes a common occurence, it would create a larger problem that would need to be remedied quickly. The same logic applies for properly flairing a post rather than leaving it to 'Other'. 

![s](http://ahs.mcnairycountyschools.com/uploads/5/2/7/9/52798193/9759366_orig.gif)

<div class="alert alert-block alert-danger">
2) <b>There are posts which fit under multiple topics.</b>
</div>

Each post can only have a single flair thus forcing users to decide which topic best represents their question or idea. However this can have several repercussions. First, the way a user decides on the topic is very subjective and may not be the best fit. Second, if they don't know which topic to pick for their post, they may decide to simply classify it under 'Other', which goes back to the first problem. 

<div class="alert alert-block alert-danger">
3) <b>The current set of available topics may not contain the best variety or be most efficient.</b> 
</div>

This is a more interesting problem and can be an analytical problem for another project because it is difficult to ascertain how efficient it is to use the current batch of flairs. One question is whether we should add more topics in order to reduce the number of 'Other' posts and to also make it easier for users to identify the proper flair for their submissions. For example, should we include a new 'Student' topic for questions pertaining to student loans, textbook savings, getting a campus job, etc. Another question is whether or not we can *remove* any of the current topics and replace it with something else. This is not to say that the current topics are bad; in fact, using an eye-test would be enough for most people to be satisfied with them but how do we know whether or not they are the most efficient set of topics? 

To sum up these 3 main problems into a single question: 

<div class="alert alert-block alert-info">
<b>How can we create an efficient set of topics and automatically assign a single one for each post?</b>
</div>

---------------------------------

>### The Project

<div class="alert alert-block alert-success">
<b>The main goal of this project is to create a topic-modeling algorithm that will generate a new set of topics for /r/personalfinance posts.</b>
</div>

This will consist of 4 stages to accomplish this task.

**1) [Data Acquisition](#1.-Data-Acquisition)**: This is often the most important step in any data science project (along with cleaning the data). We will first need to acquire the relevant data from /r/personalfinance using pushshift.io's API as well as Reddit's official API. 

**2) [Data Wrangling](#2.-Data-Wrangling)**: After we extract the data, we will need to clean it and prepare it for analysis. Natural Language Processing projects require pre-processing the text so that the machine readily able to interpret it after vectorization. Any incoherent symbols or punctuation will make the model performance worse so this is crucial before creating the model.

**3) [Exploratory Data Analysis](#3.-Exploratory-Data-Analysis)**: We will begin to explore the data and observe any interesting trends that can provide a direction or foundation for how we proceed with the project. Natural Language Processing projects are typically limited in how much EDA can be performed because of the lack of features. After all, we are mainly dealing with text!

**4) [Machine Learning](#4.-Machine-Learning)**: Let us split the problem statement into 2 parts: 

<div class="alert alert-block alert-danger">
<b>First, how can we create topics based on the posts' text and assign one to each post?</b>
</div>

- The main type of algorithm we seek to use is a topic model, which discovers topics based on hidden semantic structures in the various documents of a corpus (which is a large unstructured set of text). More specifically, this will require the implementation of Latent Dirichlet Allocation, a generative statistical model that works well in topic modeling by mapping each document to a number of different topics and each topic to a number of different words. To evaluate how well a generated set of topics does in clustering the documents, we will be using 2 metrics: Perplexity and Coherence (this will be discussed in greater detail in the machine learning section).

<div class="alert alert-block alert-danger">
<b>Second, how do we evaluate whether or not these topics are efficient?</b>
</div>

- To answer this question, we must think about how we can define an "efficient" topic. Remember that we want this model to be able to not only do well in creating topics for the fitted data, but also do a good job in assigning topics for *new* posts. The problem we face is that there are many incidents where a user does not select a topic for their own posts. Therefore this model should help resolve this issue by automatically creating a topic for these new posts. 


- The best way to evaluate the efficiency of the new topics that we generate from our topic model is by measuring the accuracy of predicting the labels. In other words, we will be creating a new classification problem where the feature variables are the text for each post and the response variable is the generated topic for each post. We then use different machine learning algorithms (such as Naive Bayes and Stochastic Gradient Descent) to create a good model that can predict the topics for a given row. Finally, we use the Accuracy, Precision and Recall metrics to evaluate how well the model performed.

***
<div class="alert alert-block alert-warning">
Note: We did not use inferential statistics to do hypothesis testing or to observe relationships between the feature and response variables. This is because our feature variables are the vectorized form of words and there are thousands of different words in our usable dictionary, which would make any statistical insight practically meaningless.
</div>

***

>### The Clients

1) **/r/PersonalFinance Users** will be interested in having a new system that could automatically determine a proper topic for their submission so that they don't have to choose one themselves. Whether they are too lazy, don't know which topic to select This will also help other people who search posts by topic. For example, if someone was interested in reading advice for people struggling with low income, they could simply navigate through the 'Planning' topic. 

![s](https://watermarked.cutcaster.com/cutcaster-photo-100747197-Cartoon-boy-working-with-computer.jpg)
 
 
2) **/r/PersonalFinance Moderators** will benefit from having a clearer structure for all submissions. Many of the popular subreddits insist on their members tagging their posts with flairs because of the convenience and organization it provides to the subreddit as a whole (sometimes this is even a requirement). One of the problems that moderators face with their subreddits is the influx of re-occuring questions. Many subreddits will ask users to first search for similar posts before submitting their own because it helps to reduce the number of the same posed questions. By having a better topic system, this will help the search feature because there will be fewer 'Other' posts and more of the specific topics. 
 
 ***

## 1. Data Acquisition

[Return to Table of Contents](#Table-of-Contents)

>**Obtaining the data**

For this report, I created my own dataset through 2 steps:

**First**, we extract the submission information using the __[pushshift.io Reddit API](https://github.com/pushshift/api)__. It gives us the ability to search through Reddit data and create data aggregations. Using sqlite was necessary as there are many different rows of data to collect.

We want to first specify the date ranges from which we can extract the data. This is important so that we can attach the date ranges to the pushshift url. For example, if we wanted to extract the newest post submissions as of August 25th 2018 for /r/nba (a subreddit for fans of the National Basketball Association), we would first convert August 25th 2018 11:59:59 PM into datetime format and then into a timestamp using the .timetuple() method. We then add the subreddit and timestamp as strings in the url. 

This would be the resulting url: http://api.pushshift.io/reddit/search/submission/?subreddit=personalfinance&size=500&before=1535255999

Take a quick look!

![s](https://github.com/nysportsfan/Personal_Finance_Subreddit/blob/master/Images/Pushshift%20IO%20Example.png?raw=true)

Notice that this will provide the 500 most recent posts as of August 25th 11:59PM because we specified size = 500. We then use the requests module to use its builtin JSON decoder and extract the relevant data from each pull. To get the next 500 most recent posts, we simply do another pull by replacing the timestamp such that it comes right after the date and time of the 500th post. 

**Second**, I used the __[official Reddit API](https://praw.readthedocs.io/en/latest/)__ to extract information that is not provided through the pushshift.io Reddit API. In this case, we extracted the flair (which will act as the "topics" of each submission) and self-text information (which are the body of text for each submission). The reason we cannot simply use the official Reddit API to extract all the submission data is because they limit the number of submissions (1000) you are allowed to extract as well as the date range. For example, if we extract the 1000 most recent posts on a subreddit, we cannot extract the next 1000 most recent posts since we cannot specify the date range. We will have a dataset that contains approximately 100,000 rows of data. Each row represents a submission in the Personal Finance subreddit. The features will include:

- title

- date (the date at which the post was submitted)

- time (the time at which the post was submitted)

- upvotes (number of upvotes for each submission)

- id (submission ID)

- topic (flair)

- self-text (the text information within each post)

To see the code and more annotation, check out the [Data Acquisition](https://github.com/nysportsfan/Personal_Finance_Subreddit/blob/master/01_pf_data_wrangling.ipynb) notebook.

<div class="alert alert-block alert-warning">
Note: This process of scraping data is very computationally expensive. Consider using PaperSpace or Google Cloud to speed up the process.
</div>

***

## 2. Data Wrangling

[Return to Table of Contents](#Table-of-Contents)

>**Inspecting the Dataset**

Let's begin by observing the dataset.

In [1]:
import pandas as pd
from pf_files import *

df = pd.read_pickle(r'C:\Users\Joshua\Pickle_files\df')
first_df = df[['title','date','time','upvotes','id','topic','self_text']]

first_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23182 entries, 0 to 23181
Data columns (total 7 columns):
title        23182 non-null object
date         23182 non-null object
time         23182 non-null object
upvotes      23182 non-null int64
id           23182 non-null object
topic        23182 non-null object
self_text    23182 non-null object
dtypes: int64(1), object(6)
memory usage: 1.2+ MB


In [2]:
first_df.head()

title        date      time  \
0                     Ways to make extra side money?  2018-09-19  12:57 PM   
1  (Year UPDATE) Legally blind, going homeless, h...  2018-09-19  12:56 AM   
2                               19, being kicked out  2018-09-19  12:55 PM   
3                            Online Savings Account?  2018-09-19  12:54 PM   
4            Tools for Managing Incomes and Expenses  2018-09-19  12:52 PM   

   upvotes      id    topic                                          self_text  
0        1  9h6whn  unknown                                                     
1       16  9h29g7    Other                                                     
2        2  9h6vyv    Other  So i just found out last night the home ive be...  
3        1  9h6vs4   Saving  Hello! Looking for recommendations for an onli...  
4        0  9h6v48    Other

>**Creating the Text column**

Rather than having the title and the self-text in separate columns, it would be more convenient to have both in a single column. We do so by concatenating the 2 columns into a new column called 'text'.

In [3]:
df[['title','self_text','text']].head()

title  \
0                     Ways to make extra side money?   
1  (Year UPDATE) Legally blind, going homeless, h...   
2                               19, being kicked out   
3                            Online Savings Account?   
4            Tools for Managing Incomes and Expenses   

                                           self_text  \
0                                                      
1                                                      
2  So i just found out last night the home ive be...   
3  Hello! Looking for recommendations for an onli...   
4                                                      

                                                text  
0                    Ways to make extra side money?   
1  (Year UPDATE) Legally blind, going homeless, h...  
2  19, being kicked out So i just found out last ...  
3  Online Savings Account? Hello! Looking for rec...  
4           Tools for Managing Incomes and Expenses

>**Convert Time and Dates into datetime format**

Since the data is given as a string, we will need to convert the time and date into datatime format.

In [4]:
df[['date','time']].head()

date      time
0  2018-09-19  12:57 PM
1  2018-09-19  12:56 AM
2  2018-09-19  12:55 PM
3  2018-09-19  12:54 PM
4  2018-09-19  12:52 PM

>**Replace missing topics**

Some posts have missing topics and self-text. This is because the post was either deleted by the user or removed by a moderator. Since we plan on creating our own topics, it is not a big deal if we are missing the topic. Let's fill missing topics by categorizing them under 'unknown'. 

In [33]:
df1 = pd.read_csv(r'C:\Users\Joshua\Downloads\Data\reddit\reddit_pf.csv', engine = 'python', index_col=0)

df1.fillna('unknown').topic.value_counts()

Debt                 1256
Other                1205
Credit               1048
Investing             859
Retirement            824
Employment            724
Housing               709
unknown               651
Auto                  560
Planning              526
Saving                524
Taxes                 509
Budgeting             405
Insurance             380
Meta                    2
THIS IS A SPAMMER       1
Name: topic, dtype: int64

>**Replace outlier topics**

As we can see from the previous point, there are two outlier topics: 2 instances of 'Meta' and 1 instance of 'THIS IS A SPAMMER'. Since these are clearly not intended to be part of the set of topics, we should replace these with 'unknown'.

In [5]:
df.topic.value_counts()

Debt          2819
Other         2749
Credit        2447
Investing     1910
Retirement    1814
Employment    1688
Housing       1574
unknown       1396
Auto          1303
Planning      1229
Saving        1200
Taxes         1164
Budgeting     1012
Insurance      877
Name: topic, dtype: int64

>**Text pre-processing**

As the saying goes: __["Garbage In, Garbage Out"](https://en.wikipedia.org/wiki/Garbage_in,_garbage_out)__. Before tokenizing the text and feeding it into the machine, we will need to clean up the data. To do so, we will create a function that will perform the following steps:

- **Lowercase** the words so that the model will not differentiate capitalized words from other words.

- **Remove numbers/digits** since the model is interpreting *text* not numbers.

- **Remove punctuation** since it is not important for the context.

- **Strip white space** since empty strings could be interpreted as text and we want to avoid that.

- **Remove stopwords**, which are general words that are very frequent in the English dictionary (ex. because, such, so). Here is a __[list of some common stopwords](https://www.ranks.nl/stopwords)__.

- **Remove noise** that is not picked up through the other cleaning methods. This step can come either before or after tokenization and normalization, or both (ex. dropping words that are less than 2 characters long).

In [6]:
df[['text','clean_text']].head()

text  \
0                    Ways to make extra side money?    
1  (Year UPDATE) Legally blind, going homeless, h...   
2  19, being kicked out So i just found out last ...   
3  Online Savings Account? Hello! Looking for rec...   
4           Tools for Managing Incomes and Expenses    

                                          clean_text  
0                         ways make extra side money  
1  year update legally blind going homeless one j...  
2  kicked found last night home ive staying going...  
3  online savings account hello looking recommend...  
4                    tools managing incomes expenses

>**Check for missing text**

After pre-processing the text, there are some cases where it completely removes all text. We will remove these rows from the dataset.

In [36]:
df.loc[[2845, 3646, 4592, 4705], ['text', 'clean_text']]

text clean_text
2845  1099 or W2?            
3646       401Ks?            
4592         401k            
4705            T

>**Setting a word limit**

We also limited the number of words in the text since the longer the text, the longer the computation time. We did consider using the number of characters but the number of words seemed to do a better job. 150 was used as the word limit for each row.

In [7]:
txt_wrd_mean = df['clean_text'].apply(lambda x: x.split()).map(len).mean()
txt_wrd_median = df['clean_text'].apply(lambda x: x.split()).map(len).median()

print('Mean Word Length: {}\n\nMedian Word length: {}\n'.format(txt_wrd_mean, txt_wrd_median))

word_count_ = len(df[df['clean_text'].apply(lambda x: x.split()).map(len)<150])
total_len = len(df)

print('Percentage of texts with word count less than 150: {0:.2f}%'.format((word_count_/total_len)*100))

Mean Word Length: 66.79087222845311

Median Word length: 49.0

Percentage of texts with word count less than 150: 89.47%


>**Tokenization**

In order to better analyze individual words, we will need to *tokenize* the documents (or in this case, the submission titles) into pieces of words. By doing so, we will be able to use the various NLP libraries to further dissect the tokens (ex. spaCy).

>**Lemmatization**

After tokenizing the data, we will need to normalize the text through lemmatization or stemming. Lemmatization is typically a better method since it returns the canonical forms based on a word's lemma. However, this process takes much more time compared to stemming the words, which simply removes the affixes of a given word.

In [8]:
df[['text','clean_text', 'lemmatized_text']].head()

text  \
0                    Ways to make extra side money?    
1  (Year UPDATE) Legally blind, going homeless, h...   
2  19, being kicked out So i just found out last ...   
3  Online Savings Account? Hello! Looking for rec...   
4           Tools for Managing Incomes and Expenses    

                                          clean_text  \
0                         ways make extra side money   
1  year update legally blind going homeless one j...   
2  kicked found last night home ive staying going...   
3  online savings account hello looking recommend...   
4                    tools managing incomes expenses   

                                     lemmatized_text  
0                    [way, make, extra, side, money]  
1  [year, update, legally, blind, go, homeless, o...  
2  [kick, find, last, night, home, have, stay, go...  
3  [online, saving, account, hello, look, recomme...  
4                    [tool, manage, income, expense]

>**Named Entity Recognition**

Named entities are real-world objects that have a name, such as a person, country, or company. spaCy is able to recognize different types of named entities in a document and can return features such as the label (ex. ORG - organization, GPE - geopolitical entity). Not all documents will contain named entities since they may not reference any. While this is not a feature we plan to incorporate in the topic model, it will be interesting for observation in the exploratory data analysis.

In [9]:
df[['named_entities', 'entity_labels']].head()

named_entities entity_labels
0                                 []            []
1                              [one]    [CARDINAL]
2  [last night, returning make week]  [TIME, DATE]
3                            [first]     [ORDINAL]
4                                 []            []

## 3. Exploratory Data Analysis

[Return to Table of Contents](#Table-of-Contents)

The goal of this notebook is to explore the data extracted from Reddit's Personal Finance subreddit, including information such as:

- Text of submission

- Date of submission

- Topic

- Number of Upvotes

- Length of text

Using Natural Language Processing tools and data visualization, we aim to learn any obvious and underlying trends from the submission information. However, it should be worth mentioning that there won't that many details to uncover compared to other types of datasets due to the fact that the model's only input will be text. Therefore many of the following findings may not be relevant to creating the topics but nonetheless, they are still interesting to see.

>**Which are the most common topics?**

Debt is the most common topic in the personal finance subreddit, followed by Other and Credit. This indicates that debt is a major concern for many of the redditors (users) and they make submissions in order to seek advice.

As discussed before, Other would include posts that either haven't been given a topic or didn't match the main topics. With more data, it could be possible that Other is the most frequent topic instead of Debt.

Investment and Retirement are also similar in their counts while also being very similar in their functions.

![s](https://github.com/nysportsfan/Personal_Finance_Subreddit/blob/master/Images/PersonalFinance%20Topic%20Count.png?raw=true)

> **Which are the top 10 most popular submissions?**

By popular, we mean which submissions had the most user upvotes. 

There are a variety of different subjects in the top 10 most popular posts, ranging from scams to dealing with the loss of a family member. However, 3 of the top 10 include the word "scam", showing that this may be a popular subject in /r/personalfinance and could happen often. This could be a common issue among many people living in the US based on the number of upvotes.

While some posts are seeking advice, others provide advice such as '"Hidden" costs of buying a home and how to prepare for them.'

In [10]:
# Top 10 most popular headlines
[title for title in df.sort_values(by='upvotes', ascending=False)['title'].values[:10]]

["IRS will allow employers to match their employees' student loan repayments",
 "If you can't get your emergency fund to grow because of emergencies that keep coming up, you're still doing a good job.",
 'Your amazon store card is probably scamming you',
 'A story of how I just got out of paying a $1500 bill, why you should NEVER blindly trust a Debt collection agency, and ALWAYS request proof of a Debt owed. 😂😂😂',
 'Girlfriend had some standard medical tests done. The clinic apparently waited too long to bill her insurance company and insurance is declining to pay them. Now, the clinic is trying to bill girlfriend for full amount and threatening to go to collections. What are her options?',
 "If the only reason you pay for Amazon Prime is because of 2-day shipping, there is a good chance it's a smarter financial move for you to cancel it than continue.",
 'So I fell for a scam yesterday and it still angers me.',
 'Credit freezes are now free. Starting today.',
 "I'm 18 and I 

> **Which are the most common tokenized words?**

In the plot below, we can see the frequency of the top 100 most common tokens with ticks on intervals of 10.

The word 'pay' has been used over 20,000 times in the corpus of text. After the first 10 tokens, the frequency that a word is used decreases very quickly. The words 'account' and 'payment' are also fairly common and transferable among many different topics which is why it is logical for them to have high counts. Towards the end of the graph, we can start to see more topic-specific words such as 'rent' and 'invest'.

![s](https://github.com/nysportsfan/Personal_Finance_Subreddit/blob/master/Images/PersonalFinance%20Token%20Count.png?raw=true)

> **What does the distribution of the text length look like?**

As we can see, the histogram of the distribution of the text length is abnormally right-skewed with a strong kurtosis. While the graph peaks around 75-100 characters, it is difficult to tell where the mean of the data is if we only look at the graph. In reality, the mean number of characters for the posts is 602, which is far from the peak of the data. There is a strong peak but after about 150 words is a very flat-lined distribution where it gradually increases up until 600 and then begins to decline. 

One interpretation of this graph can be that there are many new users who simply want to ask a quick question without going into much detail. For example, a person may want to ask about whether they should get a Visa or MasterCard credit card. These types of posts are very common where people do not have to adequately describe their financial situation to get advice.

Another interpretation is that many of the posts get deleted, resulting in the self-text disappearing and significantly reducing the number of characters in the total text. This is certainly plausible because of how many subscribers /r/personalfinance has. The more popular a subreddit becomes, the more common it is for new users to disregard the subreddit's rules and consequently, a higher chance that a moderator removes their post. In fact, the mean number of characters for posts under the 'unknown' topic (which is for posts that have been deleted or removed) is 130, which is well below the average of 602.

![s](https://github.com/nysportsfan/Personal_Finance_Subreddit/blob/master/Images/PersonalFinance%20Text%20Length%20Distribution.png?raw=true)

While this plot describes how the distribution is like for all posts, let's also take a look at the distribution by topic.

We can see depending on the topic, the kurtosis of the histogram can vary greatly. For example, the histograms for 'Other' and 'unknown' have much heavier weight on their tails compared to the other topics such as 'Budgeting', which is appears very flat. Besides their peaks, the histograms look very similar to one another which means that outside of 'unknown', the topics have very similar distributions. 

![s](https://github.com/nysportsfan/Personal_Finance_Subreddit/blob/master/Images/PersonalFinance%20Text%20Length%20Distribution1.png?raw=true)

>**Most popular tokens by POS (Part-of-Speech) Tagging**

Part of Speech tagging is marking a word based on its __[part of speech](https://en.wikipedia.org/wiki/Part_of_speech)__. Examples can include noun, verb, and pronoun. It was initially done by hand (ex. [Brown Corpus](https://en.wikipedia.org/wiki/Brown_Corpus)) but there are now many different algorithms that can be used to decipher a POS tag for any given document. spaCy is a python library that has one such algorithm that allows us to obtain the POS for a given word. Let's take advantage of this by extracting singular and plural nouns from each document. 

![s](https://github.com/nysportsfan/Personal_Finance_Subreddit/blob/master/Images/PersonalFinance%20Noun%20Token%20Count.png?raw=true)

We see that the most common token is 'credit', not 'pay' like we had seen before. This hints at the fact that 'pay' is mainly used as a verb rather than a noun. Another interesting finding is that there are many date-based words such as 'year' and 'month'. Most likely this is because many people are describing their financial circumstances in the context of a year or several months. Other common tokens strongly suggest which topics they're part of. For example: 

- $'car' -> Auto$

- $'debt' -> Debt$

- $'job' -> Employment$

- $'loan' -> Debt$

>**What are the most common named entities?**

By far the most common types of named entity are time and date entities. This indicates that most users place a great amount of detail on the context of time when they are describing their financial issues. 

![s](https://github.com/nysportsfan/Personal_Finance_Subreddit/blob/master/Images/PersonalFinance%20NER.png?raw=true)

>**Sentiment Analysis**

While observing the relative sentiments of tokens is not particularly useful in determining the topics of the posts, it is nonetheless interesting to observe due to some surprising (and humorous) results. We can apply it to our text through the SentimentIntensityAnalyzer function from nltk's sentiment module.

'Credit' is easily the most common positive word from the dataset, but this does beg the queston of whether or not it's truly positive in the context of the text. For example, what if they are referring to credit cards; would the word 'credit' still be considered positive? Similarly, the word 'interest' is the second most common positive word but interest in the financial context is probably negative! 

![s](https://github.com/nysportsfan/Personal_Finance_Subreddit/blob/master/Images/PersonalFinance%20Positive.png?raw=true)

On the flip side, we see that 'debt' is the most common negative word and this isn't surprising since it's also the most popular topic on /r/personalfinance. Whether or not the word 'pay' is necessarily negative can be debatable in this context since it could simply be neutral. For example, if someone has to "pay" for a cup of coffee, it's not necessarily negative since they are making a transaction to buy a drink for consumption. 

![s](https://github.com/nysportsfan/Personal_Finance_Subreddit/blob/master/Images/PersonalFinance%20Negative.png?raw=true)

***

## 4. Machine Learning

[Return to Table of Contents](#Table-of-Contents)

As discussed before, there are two main parts for this topic-modeling project:

1) The first part will be to generate the topics using Latent Dirichlet Allocation and then labeling each post with a main topic based on their topic distributions. This is an unsupervised task.

2) The second part will be to predict the topic for each post using various algorithms and then to evaluate their efficiency through the accuracy, precision and recall metrics. This is a supervised classification task.

Without further ado, let's jump right into it!

> ### Part 1: Creating the Topics

Like any other machine learning project, we begin by first creating features that we will be inputting into the LDA model. We do this by using a bag-of-words model which assigns each token an id number and also counts the frequency of a token in each text (document). To implement this feature with gensim (a popular library for NLP purposes), we fit the corpus of text into a dictionary. This dictionary then converts all of the tokens into a bag-of-words model.

For example, take a look at the first 10 tokens in our dictionary:

In [60]:
count = 0
for x in dictionary.token2id.items():
    count +=1
    print(x)
    if count == 10:
        break

('extra', 0)
('make', 1)
('money', 2)
('side', 3)
('way', 4)
('blind', 5)
('go', 6)
('help', 7)
('homeless', 8)
('job', 9)


Notice how each word is assigned a number next to it. That is its token ID for vector representation. This is done because the machine doesn't interpret documents as a collection of words since it is unable to understand the definition of a word. However, it **is** able to interpret words as a vector. Now let's take a look at the bag-of-words representation of the first document in the corpus: "way make extra side money".

In [61]:
# Create a bag of words
corpus = [dictionary.doc2bow(text) for text in data_bigrams]
# with open('pf_corpus.pickle', 'wb') as f:
#     pickle.dump(corpus, f)

# View sample text and its respective bag of words transformation
print(data_bigrams[0])
print(corpus[0])

['way', 'make', 'extra', 'side', 'money']
[(0, 1), (1, 1), (2, 1), (3, 1), (4, 1)]


You can see that the tuple underneath each token of the document contains the token ID and a second number. That second number represents the frequency of that particular token in the document. For example, the token 'way' has the number 1 as its frequency. This means that there is only one instance of the token 'way' in the entire document.

Now that we have created a dictionary that assigned a number to each token and created a bag-of-words model for each document, let's begin creating the Latent Dirichlet Allocation model. 

Here's how it works (click **[HERE](#Creating-the-Model)** if you want to skip the explanation of the model).

> #### How LDA Works

A corpus can be represented in vector space by a document-word matrix which contains the frequency of every word in a corpus against every document in the corpus. Here is an example:

![s](https://www.researchgate.net/profile/Mark_Cutkosky/publication/220677827/figure/fig2/AS:394013053079585@1470951440970/ector-space-representation-of-a-document-collection-or-corpus-matrix.png)

The value $c\scriptsize11$ represents the number of times the word 'schedule' appears in the document 'Memo'. So how does this relate to creating the topics? What LDA does is essentially split this $M x N$ matrix into 2 lower dimensional matrices, which we will call $A1$ and $A2$. $A1$ is a document-topics matrix whereas $A2$ is a topic-word matrix. Let the number of topics be $K$; $A1$ is therefore a $M x K$ matrix while $A2$ is a $K x N$ matrix. 

There are now 2 major steps in how it assigns a topic to each document:

1) The model first *randomly* assigns a set of $K$ topics to each document. In other words, these topics can be completely wrong in how well it represents the clusters of documents. You can think of them as dummy topics at this point in time.

2) Each topic generates words based on their probability distribution. For example, if 'dog' appears more frequently than other words in documents with topic 1, then 'dog' will have a strong weight in topic 1. By repeatedly iterating through the different words in each document, the model attempts to adjust itself by assigning a new word to a topic. If the model discovers that the word 'dog' is infrequent in topic 2, then it will correct itself by replacing 'dog' with another word with a higher probability $P$.

The way we calculate $P$ is through multiplying 2 smaller probabilities:

1) $P(topic\;t|document\;d)$ - the proportion of words in document $d$ that are currently assigned to topic $t$.

2) $P(word\;w|topic\;t)$ - the proportion of assignments to topic $t$ over all documents that come from the word $w$.

In other words, $P$ is the probability that topic $t$ generated word $w$. 

The figure below describes how LDA functions in a nutshell. Each topic has a discrete probability distribution of words and each document is composed of a mixture of topics. 

![s](http://bigdata.ices.utexas.edu/wp-content/uploads/2015/01/LDA-concept.png)

> #### Creating the Model

Actually creating the model using gensim's LDA model is a relatively straight-forward process because we only need to input 2 parameters into gensim's LDA model, the corpus of text and the dictionary. However, in order to optimize the model there are additional parameters that we need to focus on:

1) **number of topics (num_topics)** - The number of topics that are created from the bag-of-words

2) **alpha value (alpha)** - The alpha value controls the mixture of topics for any given document. The greater the alpha, the greater the variety of topics will appear for any given document. The lower the alpha, the lower the variety of topics.

According to Griffiths and Steyvers (2004), it is suggested to use a value of 50/k (where k is number of topics) for α. However, gensim allows us to specify 'auto' for the alpha parameter, which causes the machine to automatically estimate alpha based off of the corpus we train with. This conveniently means that we won't have to test out different alpha values.

To determine which are the best hyper-parameters, we will be testing different number of topics and evaluating each iteration of the model based on its performance on its **coherence value**. Coherence is the measure of how much a document makes sense given the context of the words in relation to each other. For example, a coherence score of 0.10 would indicate that the model did a poor job at creating topics for similar documents.

Let's take a look at the results when we use gensim's built in LDA wrapper.

![s](https://github.com/nysportsfan/Personal_Finance_Subreddit/blob/master/Images/PersonalFinance%20Coherence%20Graph.png?raw=true)

Coherence score peaks when the number of topics is equal to 20. This means that the LDA model produces the most coherent and best variety of clustered documents when there are 20 topics. Therefore this would be the most "optimal" number of topics. Let's also use the __[pyLDAvis](https://github.com/cpsievert/LDAvis)__ package to display how the topics vary in relation to each other.

In [4]:
corpus = [dictionary.doc2bow(text) for text in data_bigrams]
optimal_model = model_list[3]
import pyLDAvis
import pyLDAvis.gensim 
import gensim
pyLDAvis.enable_notebook()
vis = pyLDAvis.gensim.prepare(optimal_model, corpus, dictionary)
vis

C:\Users\Joshua\Anaconda3\lib\site-packages\pyLDAvis\_prepare.py:257: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  return pd.concat([default_term_info] + list(topic_dfs))


PreparedData(topic_coordinates=              x         y  topics  cluster       Freq
topic                                                
16    -0.038368  0.088166       1        1  12.610233
5      0.045948 -0.084265       2        1  10.030854
18    -0.066907 -0.043867       3        1   9.720840
3      0.017819  0.117111       4        1   8.196844
4     -0.028707 -0.045783       5        1   7.745319
10     0.014935  0.182867       6        1   6.405858
8     -0.078295 -0.010573       7        1   5.481642
2     -0.018336 -0.077663       8        1   4.430221
12    -0.006647 -0.049893       9        1   4.393626
0     -0.015330 -0.024978      10        1   4.199517
11    -0.101606 -0.045300      11        1   4.092866
9      0.181794 -0.115881      12        1   3.934385
15     0.083307  0.058827      13        1   3.750660
14    -0.135852 -0.112141      14        1   3.000365
1      0.025933  0.010571      15        1   2.914912
6      0.217205 -0.010559      16        1   2.842705
19     0.020366  0.022874      17        1   2.184103
17    -0.103471  0.048470      18        1   1.871426
13    -0.025390  0.079738      19        1   1.196877
7      0.011601  0.012278      20        1   0.996746, topic_info=     Category          Freq        Term         Total  loglift  logprob
term                                                                   
117   Default  15392.000000      credit  15392.000000  30.0000  30.0000
19    Default  10050.000000         car  10050.000000  29.0000  29.0000
114   Default  10160.000000        card  10160.000000  28.0000  28.0000
169   Default  11825.000000        loan  11825.000000  27.0000  27.0000
466   Default   7098.000000         tax   7098.000000  26.0000  26.0000
377   Default   5619.000000        fund   5619.000000  25.0000  25.0000
453   Default   3940.000000         ira   3940.000000  24.0000  24.0000
165   Default   5014.000000   insurance   5014.000000  23.0000  23.0000
51    Default  10160.000000     account  10160.000000  22.0000  22.0000
461   Default   2890.000000        roth   2890.000000  21.0000  21.0000
9     Default   9214.000000         job   9214.000000  20.0000  20.0000
73    Default  20564.000000         pay  20564.000000  19.0000  19.0000
301   Default   2841.000000        cash   2841.000000  18.0000  18.0000
126   Default   7499.000000     payment   7499.000000  17.0000  17.0000
50    Default  10462.000000        work  10462.000000  16.0000  16.0000
11    Default  13434.000000       month  13434.000000  15.0000  15.0000
244   Default   5043.000000     student   5043.000000  14.0000  14.0000
58    Default   8692.000000        debt   8692.000000  13.0000  13.0000
221   Default   5241.000000       house   5241.000000  12.0000  12.0000
132   Default   3422.000000       score   3422.000000  11.0000  11.0000
179   Default  22180.000000         not  22180.000000  10.0000  10.0000
300   Default   5341.000000         buy   5341.000000   9.0000   9.0000
77    Default   6162.000000      saving   6162.000000   8.0000   8.0000
517   Default   5202.000000     company   5202.000000   7.0000   7.0000
1304  Default   2861.000000    mortgage   2861.000000   6.0000   6.0000
210   Default   2343.000000      budget   2343.000000   5.0000   5.0000
1013  Default   2201.000000       stock   2201.000000   4.0000   4.0000
107   Default   2784.000000  retirement   2784.000000   3.0000   3.0000
238   Default   2903.000000      school   2903.000000   2.0000   2.0000
135   Default   6469.000000         use   6469.000000   1.0000   1.0000
...       ...           ...         ...           ...      ...      ...
301   Topic20    734.551453        cash   2841.802246   3.2555  -3.0235
4697  Topic20     63.730396        wire    105.481033   4.1046  -5.4681
2785  Topic20     39.069263    shipping     53.629051   4.2917  -5.9574
1514  Topic20     55.227497     english     91.402969   4.1046  -5.6113
2452  Topic20     42.227470       acorn     62.387562   4.2181  -5.8797
1250  Topic20     84.545464  calc

While there are noticeably distinct topics such as topic 16 (investing), topic 8 (401k), topic 11(employment), and topic 2 (credit card), you can see that there are many clusters around topic 1. A healthy set of topics will produce a plot with topic bubbles that are fairly distant from one another and having medium-sized bubbles. Therefore this cluster could definitely use some improvement.

To do this, we will also use the __[MALLET](http://mallet.cs.umass.edu/)__ package because it often does a better job at generating topics for this very purpose.

<div class="alert alert-block alert-warning">
Note: There is currently an error when using the pyLDAvis package with LDA Mallet wrappers so we are unable to visualize the Mallet topics. 
</div>

![s](https://github.com/nysportsfan/Personal_Finance_Subreddit/blob/master/Images/PersonalFinance%20Coherence%20Graph1.png?raw=true)

As we can see, the coherence value peaks when the number of topics $K = 8$. However, when looking at the topics and doing an eye-test, it was determined that using $K = 20$ led to a better spread and variety of topics. Let's take a look below.

In [11]:
from pprint import pprint
MALLET_optimal_model = MALLET_model_list[3]
pprint(MALLET_optimal_model.print_topics(num_words=10))

[(0,
  '0.087*"move" + 0.080*"live" + 0.040*"rent" + 0.036*"parent" + 0.029*"place" '
  '+ 0.024*"family" + 0.024*"apartment" + 0.021*"find" + 0.021*"mom" + '
  '0.019*"dad"'),
 (1,
  '0.053*"make" + 0.047*"good" + 0.046*"feel" + 0.033*"lot" + 0.033*"thing" + '
  '0.029*"financial" + 0.026*"life" + 0.026*"money" + 0.023*"pretty" + '
  '0.018*"idea"'),
 (2,
  '0.047*"call" + 0.034*"receive" + 0.026*"send" + 0.026*"charge" + '
  '0.019*"report" + 0.017*"collection" + 0.014*"email" + 0.014*"claim" + '
  '0.013*"check" + 0.013*"letter"'),
 (3,
  '0.107*"year" + 0.059*"school" + 0.054*"college" + 0.035*"start" + '
  '0.032*"student" + 0.025*"graduate" + 0.023*"work" + 0.021*"time" + '
  '0.018*"make" + 0.017*"job"'),
 (4,
  '0.140*"work" + 0.118*"job" + 0.073*"pay" + 0.069*"week" + 0.068*"time" + '
  '0.042*"hour" + 0.034*"day" + 0.030*"make" + 0.021*"full" + 0.020*"start"'),
 (5,
  '0.037*"give" + 0.024*"find" + 0.021*"edit" + 0.019*"thing" + 0.018*"guy" + '
  '0.015*"number" + 0.015*"pers

Great! It appears as though we have generated a good variety of topics but each one is made up of a distribution of words. There's no one-size-fits-all method for determining what theme to describe each LDA-generated topic but for the purposes of this project, we have decided to do an eye-test and use the different words in each topic to come up with a topic name. 

In [12]:
new_topics = {
    0: 'housing',
    1: 'other',
    2: 'collections',
    3: 'school',
    4: 'employment',
    5: 'other',
    6: 'credit card',
    7: 'auto',
    8: 'taxes',
    9: 'other',
    10: 'debt',
    11: 'investment',
    12: 'bank account',
    13: 'employment',
    14: 'budgeting',
    15: 'loans',
    16: 'other',
    17: 'insurance',
    18: 'housing',
    19: 'planning' 
}

new_topics

{0: 'housing',
 1: 'other',
 2: 'collections',
 3: 'school',
 4: 'employment',
 5: 'other',
 6: 'credit card',
 7: 'auto',
 8: 'taxes',
 9: 'other',
 10: 'debt',
 11: 'investment',
 12: 'bank account',
 13: 'employment',
 14: 'budgeting',
 15: 'loans',
 16: 'other',
 17: 'insurance',
 18: 'housing',
 19: 'planning'}

<div class="alert alert-block alert-warning">
Note: We named topics 1, 5, 9, and 16 with the name 'Other' since they did not fit into any recognizable topic. 
</div>

Now let's take a look at how the newly generated topics stack up!

In [13]:
df_main_topic.drop(['Doc_Number','Main_Topic'], axis=1)[:10]

Topic_Proportion                                           Keywords  \
0            0.0772  money, year, save, saving, put, start, plan, e...   
1            0.0718  move, live, rent, parent, place, family, apart...   
2            0.1090  work, job, pay, week, time, hour, day, make, f...   
3            0.1172  account, bank, money, check, open, transfer, f...   
4            0.0710  month, expense, spend, budget, monthly, income...   
5            0.1087  question, make, personal, finance, post, busin...   
6            0.1568  credit, card, score, balance, apply, limit, bu...   
7            0.1966  loan, pay, debt, payment, interest, student, r...   
8            0.1411  work, job, pay, week, time, hour, day, make, f...   
9            0.0770  call, receive, send, charge, report, collectio...   

                                                Text   Topic_Names  
0                    Ways to make extra side money?       planning  
1  (Year UPDATE) Legally blind, going homeless, h...       housing  
2  19, being kicked out So i just found out last ...    employment  
3  Online Savings Account? Hello! Looking for rec...  bank account  
4           Tools for Managing Incomes and Expenses      budgeting  
5  With resources like reddit, is financial consu...         other  
6  Credit hit from late payment, fee was waiver, ...   credit card  
7  Need Help budgetting/getting out of debt Long ...         loans  
8  22yo M I'm a 22 year old male, with an almost ...    employment  
9  Debt collector gave 2 hours to pay, yelled, sa...   collections

![s](https://github.com/nysportsfan/Personal_Finance_Subreddit/blob/master/Images/PersonalFinance%20Generated%20Topic%20Count.png?raw=true)

Other, Investments and Credit Cards are the 3 most common topics in terms of number of posts. 10.6% of all posts are classified under 'other', which could imply that these were bad classifications. With a coherence value of 0.4867, it certainly makes sense that certain topics are more "generalized" than others. In the original topics, 'other' was the 2nd largest topic by number of posts so it is not too far off. 

On the flip side, Budgeting, Planning, and Debt are the 3 least common topics. This is unusual since debt was the most common topic in the original topics as dictated by /r/personalfinance. This could be due to a few different reasons:

- 1) The posts that were originally classified as 'Debt' have now been classified under more specific debt categories such as Loans and Collections.

- 2) Redditors classified their posts as 'Debt' without actually being about debt. In other words, it could be the case that there are many false positives in the original classification.

- 3) Redditors classified their posts as 'Debt' but the posts had stronger semblance to other topics. For example the new generated topic 'school' could also include debt as part of student loans.

- 4) Neither the original topics nor the LDA MALLET model have done a good job of determining which posts are about the topic 'Debt'. 

Based on the topic-word similarity similarity percentages and doing an eye-test on the posts, it's clear that the topic model did a good job at assigning the appropriate tags for the main topics. However, we can see that the topics that were named 'Other' are more difficult to assign to posts. In order to observe how effective our topic model was at determining which posts belong in 'Other', let's now move on to evaluating via classification.

> ### Part 2: Evaluating/Predicting the Topics

By using our new generated topics as the target variables, we will use the text as the feature variables and try to predict the labels for new data. By measuring the accuracy, precision and recall of the predictions we will be able to determine how reasonable and logical the topics are. For example, if it turns out that many of the posts labeled as 'school' were indeed predicted as 'school', our model would have done a good job of capturing most of the school-related posts into a single cluster.

**Assumptions**: 

<div class="alert alert-block alert-warning">
We are assuming that the labels that have been given by the topic model accurately represent what the posts are actually about. However, this is almost certainly not true for all posts and because of this, we will be observing type I and type II errors along the way (particularly for posts that were labeled as 'Other' by the topic model.)
</div>

**Feature Selection**

We begin by defining our features (the cleaned text) and our response variables (our generated topics).

In [14]:
df_eval[['clean_text','new_topic']].head()

clean_text     new_topic
0                         ways make extra side money      planning
1  year update legally blind going homeless one j...       housing
2  kicked found last night home ive staying going...    employment
3  online savings account hello looking recommend...  bank account
4                    tools managing incomes expenses     budgeting

Just like we created a bag-of-words model before we applied the topic modeling techniques, we will once again have to vectorize the text. Only this time, we will be using the CountVectorizer function from sci-kit learn. While using a bag-of-words model does the job in counting the frequency of each word per document, we want to better quantify the signature words which don't appear as frequently as common words. For this reason, we will also introduce a **term frequency - inverse document frequency (tf-idf) model**. We apply the tf-idf transformer, which will transform the word count matrix in each vector into a tf-idf representation. Finally, we apply the classifier to create the machine learning model.

**Models**: 

5 different types of machine learning algorithms were implemented:

- __[Logistic Regression](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html#sklearn.linear_model.LogisticRegression)__


- __[Linear SVC](http://scikit-learn.org/stable/modules/generated/sklearn.svm.LinearSVC.html#sklearn.svm.LinearSVC)__


- __[Multinomial Naive Bayes](http://scikit-learn.org/stable/modules/generated/sklearn.naive_bayes.MultinomialNB.html#sklearn.naive_bayes.MultinomialNB)__


- __[Stochastic Gradient Descent](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDClassifier.html#sklearn.linear_model.SGDClassifier)__


- __[Random Forest](http://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html#sklearn.ensemble.RandomForestClassifier)__

We will be covering our top 2 performing models: the Logistic Regression model and the Linear Support Vector Classification model, but we will also discuss the results of all 5 models.

**Evaluation Metrics**: 

The way we will be evaluating the effectiveness of a model is through 3 metrics:

- Accuracy: Correctly labeled predictions over total predictions

- Precision: Correctly labeled true predictions over total labeled true predictions

- Recall: Correctly labeled true predictions over total actually true cases

For this project, **overall accuracy** would be the most important of the 3 because we are mainly interested in correctly labeled predictions regardless of which are the "trues" and "negatives". If it were a binary classification problem, it may be possible that precision or recall would be better but this is a multi-class classification problem thus overall precision or recall would not matter as much.

Another important evaluation metric is the **recall of the 'Other' topic** because we want to reduce the possibility that a topic that belongs in 'Other' gets predicted as a main topic. For example, if there was a post about wedding expenses, it's more than likely that it should belong in 'Other'. However, if it's misclassified by the model as 'Auto', that would be a big mistake. 

Thus, we are looking for a model whose overall accuracy and whose recall score for 'Other' are high.

**Baseline**

Since metrics like the accuracy are very relative depending on which model you use, a baseline metric should be established in order to act as a comparison to gauge how effective the model is at predicting the topic. The method chosen to create the baseline for the regressor was through the sci-kit learn function DummyClassifier. For this project, stratified was selected as the strategy.

In [2]:
import numpy as np
from sklearn.decomposition import LatentDirichletAllocation, TruncatedSVD
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer, TfidfTransformer
from sklearn.model_selection import GridSearchCV, train_test_split, RandomizedSearchCV, cross_val_score
from sklearn.linear_model import LogisticRegression, SGDClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.pipeline import Pipeline
from sklearn.svm import LinearSVC
from sklearn.multiclass import OneVsRestClassifier
from sklearn.dummy import DummyClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.naive_bayes import MultinomialNB
# Next let's select our features
X = df_eval.clean_text # The cleaned text will be the feature variables
y = df_eval.new_topic # The generated topic labels will be the response variables
topic_labels = y.unique() # List of unique topic labels

# Split the data by using 30% as the testing data 
X_train, X_test, y_train, y_test, indices_train, indices_test = train_test_split(X, y, df_eval.index, test_size=0.3, random_state=42)

# Print the accuracy 
y_pred = dummy.predict(X_test)
print('Accuracy: %s' % round(dummy.score(X_test, y_test), 4)) # You can use either the .score method or the accuracy_score function from sklearn.metrics
dummy_cv = cross_val_score(dummy,X,y,cv=10, scoring='accuracy') # Using cross-validation will provide a more accurate score
print('Average 10-Fold Cross-Validation Accuracy Score: %s' % round(np.mean(dummy_cv), 4))
print('\nClassification Report:\n')
print(classification_report(y_test, y_pred, target_names=dummy.classes_))

Accuracy: 0.0756
Average 10-Fold Cross-Validation Accuracy Score: 0.0787

Classification Report:

              precision    recall  f1-score   support

        auto       0.06      0.06      0.06       419
bank account       0.06      0.07      0.06       350
   budgeting       0.06      0.05      0.06       293
 collections       0.06      0.05      0.05       469
 credit card       0.10      0.10      0.10       642
        debt       0.02      0.02      0.02       125
  employment       0.09      0.09      0.09       593
     housing       0.11      0.11      0.11       849
   insurance       0.03      0.03      0.03       323
  investment       0.06      0.06      0.06       497
       loans       0.09      0.08      0.09       550
       other       0.10      0.11      0.10       729
    planning       0.01      0.01      0.01       253
      school       0.05      0.06      0.05       322
       taxes       0.08      0.08      0.08       541

 avg / total       0.08      0.08   

So basically as long as we get any results better than 0.0787 accuracy and 0.10 for recall for the 'Other' topic, we will have outperformed the baseline.

Let's take a look below at the Linear SVC and Logistic Regressions results in particular since they performed the best.

**Linear SVC**

In [3]:
# Extract the model with the best parameters
best_svc = gs_svc.best_estimator_
# Predict the labels
y_pred = best_svc.predict(X_test)
# Extract an array of the labels
x = best_svc.classes_
# Obtain the cross-validation scores for 10-folds using accuracy as the scoring metric
svc_cv = cross_val_score(best_svc,X,y,cv=10, scoring='accuracy')

# Print the accuracy
print('Accuracy: {0:.4f}'.format(best_svc.score(X_test, y_test)))
print('Average 10-Fold Cross-Validation Accuracy Score: %s' % round(np.mean(svc_cv), 4))
print('\nClassification Report:\n')
print(classification_report(y_test, y_pred, target_names=x))

Accuracy: 0.7896
Average 10-Fold Cross-Validation Accuracy Score: 0.7945

Classification Report:

              precision    recall  f1-score   support

        auto       0.87      0.88      0.88       419
bank account       0.79      0.83      0.81       350
   budgeting       0.80      0.74      0.77       293
 collections       0.80      0.74      0.77       469
 credit card       0.89      0.90      0.90       642
        debt       0.59      0.35      0.44       125
  employment       0.80      0.83      0.82       593
     housing       0.77      0.82      0.79       849
   insurance       0.81      0.82      0.81       323
  investment       0.84      0.85      0.85       497
       loans       0.82      0.83      0.83       550
       other       0.61      0.60      0.60       729
    planning       0.70      0.58      0.63       253
      school       0.76      0.78      0.77       322
       taxes       0.83      0.88      0.85       541

 avg / total       0.79      0.79   

![s](https://github.com/nysportsfan/Personal_Finance_Subreddit/blob/master/Images/PersonalFinance%20Linear%20SVC.png?raw=true)

Using Linear Support Vector Classification resulted in an overall accuracy of 0.7945 and recall of 0.60 for 'Other'.

On paper this looks pretty good, especially compared to our baseline. If we take a look at the classification report, it's worthwhile noting that while some topics like credit card and auto had high recall rates, others like debt and other did poorly. Going back to our business-case problem, we want to know how this impacts the number of 'Other' posts. Let's take a look at how accurate using Linear SVC was in predicting 'Other' posts. 

Below are samples from the testing set which were labeled as 'Other' and correctly predicted as 'Other'.

In [106]:
pd.set_option('display.max_colwidth', 150)
other_other_list = pd.read_pickle('svc_other_other_list')
print('"Other" posts predicted as "Other"')
df_eval.iloc[sorted(other_other_list), [0]][:10]

"Other" posts predicted as "Other"


text
169                                                          I've learned that I'm marrying into extreme wealth and I'm having a hard time processing it. 
180                                                                                                                     Is this person trying to scam me? 
188  How does borrowing against a 401k, 403b, or 401a work? I've tried to do some google searches, but I haven't gotten anything very useful. I'm not e...
205                                                                                                The financial situation of a 28 year old inspired post 
222  Any good apps to keep finances in order? Sorry if his has been asked before but I’ve tried a couple apps and they seem a bit out of order, when th...
323           Question about financial benefits of marriage &amp; seeking general advice for a couple coming in to an unexpected sum of money. (Bay Area) 
410  Good with money, but dumb with money. I need advice. I am 28 and I work a lot, and I don't spend money (except in a few areas). I travel for work,...
433  On a w4, am i suppose to be putting a 1 or a 0 for a single person getting a job somewhere? The woman told me she puts 0 because she has kids but ...
457                                                                                                     How Bad is the Financial Damage, am I Salvagable? 
467                                                                                                                       A reminder for all those &lt;18

As we can see from the small sample of 10 texts, only the posts on rows 188 and 433 potentially might fit under a different topic (perhaps taxes). The other 8 posts would appropriately fit under 'Other'. While this is obviously just a small sample, it does highlight the fact that our assumption may be in danger because there are obviously a few posts which should not be part of 'Other'.

We should also take a look at the 'Other' posts which were incorrectly labeled as a different topic. First, let's view the posts where 'Other' posts were misclassified as 'Auto' posts.

In [7]:
pd.set_option('display.max_colwidth', 200)
other_other_list = pd.read_pickle('svc_other_auto_list')
print('"Other" posts predicted as "Auto"')
df_eval.iloc[sorted(other_other_list), [0]]

"Other" posts predicted as "Auto"


text
3077   Question on Duty free shopping \nI am planning on traveling overseas out of LAX(los Angeles airport) around december and want to purchase an diamond ring  on Tiffany Co. as duty free.\nBut I am af...
4345   Bought a Used Car and was charged 4000 dollars more than the listed price Hey everyone, yesterday I bought a 2010 Ford Taurus SEL, The vehicle was listed on their website for 8.6 thousand dollars ...
5598   Captial gains on commercial vehicle I just have what I hope will be a quick question regarding how captial gains works when selling a commercial vehicle.  I own a two truck dump truck company regi...
10803  Offered a white gold 10k with 1/2 diamonds in it for my Xbox one. Do I take it? So just a few hours ago I was offered a ring for my Xbox, which I was selling for 250. The guy told me he was going ...
15943                                                                                                How much money would you have to make per year to feel comfortable purchasing an entry-level luxury car? 
17664  What trading platform to use for lower volume trading? After doing a little research, I want start doing some trading and I was hoping for some advice on what platform to use. I had a gift card fo...
20507  Making 9000 euros a month in Germany here as a US expat. In the contract, 2000 of it is listed as housing and car expenses. What is the best way for me to make the most out of money without paying...
21040  Are car dealership giveaways legit? I recently got a flyer for a nearby car dealership and "won" $750. The way it works is there is a code that matches with prizes. My code matched a cash prize, a...
22673  Cars on my credit report that don't belong to me - what can I do? Per CreditKarma, everything seems fine except for two cars that are listed that don't belong to me. The two cars belong to family ...
22942                                                                                             Relative asked me be guarantor for car purchase. Said I don't have to be there in person. Is it suspicious?

As we can see, while some of the posts were incorrectly predicted as "auto", they are actually about cars and thus this highlights an important point: **the original topic model has mislabeled some of the posts as 'Other'**. This shows the fallacy of our classification assumption which was that the topic model labeled every post with an appropriate topic. However, we can see here that this is not the case since these sample posts would have been better suited for the "auto" topic. Even though our recall score was only 0.60, this is no longer a useful metric since the model could be "correcting" the labels rather than being wrong. For example, row #15943 would be a case where the Linear SVC model was considered a false prediction, but in reality it is actually correct in guessing that it's an 'Auto' post.

This could have been due to the fact that the 'other' topic had many overlapping words and phrases with other topics. Another reason could have been due to the fact that more than 1 topic was classified under 'other', resulting in a mixture of differing words that are not quite coherent when used together.

**Logistic Regression**

In [36]:
best_logreg = gs_logreg.best_estimator_
y_pred = best_logreg.predict(X_test)
x = best_logreg.classes_
logreg_cv = cross_val_score(best_logreg,X,y,cv=10, scoring='accuracy')

print('Accuracy: {0:.4f}'.format(best_logreg.score(X_test, y_test))) 
print('Average 10-Fold Cross-Validation Accuracy Score: %s' % round(np.mean(logreg_cv), 4))
print('\nClassification Report:\n')
print(classification_report(y_test, y_pred, target_names=x))

Accuracy: 0.7976
Average 10-Fold Cross-Validation Accuracy Score: 0.8015

Classification Report:

              precision    recall  f1-score   support

        auto       0.87      0.88      0.87       419
bank account       0.83      0.84      0.84       350
   budgeting       0.81      0.72      0.76       293
 collections       0.82      0.75      0.78       469
 credit card       0.88      0.92      0.90       642
        debt       0.71      0.28      0.40       125
  employment       0.81      0.84      0.83       593
     housing       0.77      0.83      0.80       849
   insurance       0.82      0.80      0.81       323
  investment       0.84      0.87      0.85       497
       loans       0.83      0.85      0.84       550
       other       0.60      0.63      0.61       729
    planning       0.76      0.58      0.66       253
      school       0.80      0.77      0.78       322
       taxes       0.83      0.89      0.86       541

 avg / total       0.80      0.80   

![s](https://github.com/nysportsfan/Personal_Finance_Subreddit/blob/master/Images/PersonalFinance%20Logistic%20Regression.png?raw=true)

The Logistic Regression model performed slightly better with an accuracy of 0.8015, overall precision of 0.80, and overall recall of 0.80. 11 out of the 15 topics had a precision score of at least 0.80 and 9 had a recall score of at least 0.80. Let's look at a few examples of how accurate the predictions performed as well as how they failed.

First we'll observe the first 10 samples from the batch of 'Other' posts that were correctly predicted as 'Other'.

In [109]:
pd.set_option('display.max_colwidth', 150)
other_other_list = pd.read_pickle('logreg_other_other_list')
print('"Other" posts predicted as "Other"')
df_eval.iloc[sorted(other_other_list), [0]][:10]

"Other" posts predicted as "Other"


text
169                                                          I've learned that I'm marrying into extreme wealth and I'm having a hard time processing it. 
180                                                                                                                     Is this person trying to scam me? 
205                                                                                                The financial situation of a 28 year old inspired post 
222  Any good apps to keep finances in order? Sorry if his has been asked before but I’ve tried a couple apps and they seem a bit out of order, when th...
323           Question about financial benefits of marriage &amp; seeking general advice for a couple coming in to an unexpected sum of money. (Bay Area) 
379                                                                                    Financial Account advice when you work for that financial company? 
410  Good with money, but dumb with money. I need advice. I am 28 and I work a lot, and I don't spend money (except in a few areas). I travel for work,...
413                                                                                        Can we all complain about our absurdly high rent for a minute? 
433  On a w4, am i suppose to be putting a 1 or a 0 for a single person getting a job somewhere? The woman told me she puts 0 because she has kids but ...
457                                                                                                     How Bad is the Financial Damage, am I Salvagable?

The same posts were predicted in the first 10 samples of the batch compared to the Linear SVC model, with the exception of rows 379 and 433. Overall, the 'other' topic did slightly better in this model compared to the Linear SVC model with an 'other' recall of 0.63. This means that there were more posts where they were predicted to be 'Other' and were actually labeled 'Other'. 

However, as seen in the Linear SVC model, this could be a very misleading statistic because the model assumption was wrong. 

Let's take a look at an example where a post that was predicted as "Loans" was actually "Other".

In [121]:
pd.set_option('display.max_colwidth', 200)
other_other_list = pd.read_pickle('logreg_other_loans_list')
print('"Other" posts predicted as "Loans"')
df_eval.iloc[sorted(other_other_list), [0]][:10]

"Other" posts predicted as "Loans"


text
188   How does borrowing against a 401k, 403b, or 401a work? I've tried to do some google searches, but I haven't gotten anything very useful. I'm not even too sure what to search for specifically. I'm ...
1071             Does the Common Reporting Standard (CRS) share information to the UK Student Loan Company? Does the Common Reporting Standard (CRS) share information to the UK Student Loan Company? Thanks
3111  Can someone please help out with the mortgage math here? Hi! Thanks in advance for your help!\n\nGot a 30-yr mortgage in March ($493K @ 4%). I just realized that I can borrow up to $50K from my 40...
4010                                                                                                                     Loaning a large amount of money for a very short amount of time to a family member. 
6421                                                                                                                                                                                  Weird debt question... 
6527                                                                                                                                                                   Parents too old and in a ton of debt. 
6896                                                                                                                                                                Whats your favorite equity loan company? 
8066  Is there a problem with 60 or 72-months if the purpose is really to leverage cash? Everything I see on auto loans says avoid longer loans. I don't get it. Is that advice assuming you're doing it t...
8404  Need advice on a good rewards, lower limit, cc to pay my monthly bills with... So first off, I'm a bartender so my money is inconsistent. I always make enough money to pay my bills by the end of t...
9995  Considering joining the military to help pay down debt Hi PF!  I've lurked for a long time, but am finally posting on a throwaway to get some feedback on an idea.\n\n&#x200B;\n\nMy wife (28) and I...

Just like with the Linear SVC model, many of the 'Other' posts that were mislabeled as a main topic turned out to actually not be part of 'Other' using an eye-test. For example, row 1071 is clearly a question about loans despite having been labeled as 'Other' by the topic model. 

In this case, we had a recall score of 0.63, which surpasses the 0.60 recall score from the previous model but at this point this metric has become meaningless due to the fallacy of our assumption that posts with the 'Other' tag were accurately flaired. The reason for this fallacy is due to the nature of how we created the topics and named them. Some of the topics contain words and aspects that are shared among other topics; this is something that occured no matter the number of topics chosen. 

> #### Model Results

Let's begin by taking a look at the accuracy results for all 5 models.

![s](https://github.com/nysportsfan/Personal_Finance_Subreddit/blob/master/Images/PersonalFinance%20Results%20Boxplot.png?raw=true)

![s](https://github.com/nysportsfan/Personal_Finance_Subreddit/blob/master/Images/PersonalFinance%20Results%20Bar%20Graph.png?raw=true)


**Pros**:

Evidently, the Logistic Regression model was the best both in the mean 10-fold CV (0.8015) and the peak accuracy score (0.8175). This means that on average, the model got over 80% of all predicted topics correct. However, the Linear SVC model is very close in accuracy as well with a mean 10-fold CV accuracy score of 0.7945. Multinomial Naive Bayes surprisingly did not perform that well considering its reputation as being one of the best for text classification purposes. Part of this may have been due to the fact that Logistic Regression models tend to work better for text classification because it is inherently a discriminative objective function as opposed to Naive Bayes, which is a generative objective function ([source](http://pages.cs.wisc.edu/~jerryzhu/cs838/LR.pdf)).

When looking at the precision and recall scores for the Logistic Regression model, we could see that many of the topics did very well with scores above 0.8 and the high 0.7s. With the exception of the 'other' topic, it seems as though gensim's LDA algorithm did a good job of creating the topics and choosing the 20 as the number of topics was the right choice. 

**Cons**:

The results for the 'Other' posts, however, were not so good due to the fact that the topic model did not correctly label which posts *should not* have been part of 'Other'. For example, the post asking "Whats your favorite equity loan company?" should have been labeled under "loans" but was instead considered an 'Other' post by the topic model. 

## 5. Conclusion

[Return to Table of Contents](#Table-of-Contents)

What we have learned from creating new topics using Latent Dirichlet Allocation and then evaluating them as a classification task is that the topics did *not* differ all too much from the original set of /r/personalfinance topics. Between the 2 topic sets, they shared 10 of the same topics. The new topics that the generated topics did not have in common with the originals were 'Bank Account', 'Collections', 'Credit Card', 'Loans', and 'School'. These topics are more specific and can be considered to be already be part of other topics. For example, 'Bank Account' could be part of 'Credit' or 'Debt' and 'Collections' could be part of 'Debt'. 

In [34]:
df3 = pd.DataFrame(index=range(15))
df3['Original Topics'] = pd.Series(sorted(df_eval['old_topic'].value_counts().index)[:-1])
df3['New Topics'] = pd.Series(sorted(df_eval['new_topic'].value_counts().index))
df3.fillna('')

Original Topics    New Topics
0             Auto          auto
1        Budgeting  bank account
2           Credit     budgeting
3             Debt   collections
4       Employment   credit card
5          Housing          debt
6        Insurance    employment
7        Investing       housing
8            Other     insurance
9         Planning    investment
10      Retirement         loans
11          Saving         other
12           Taxes      planning
13                        school
14                         taxes

We also learned that topic modeling is very much based on using an **eye-test** rather than solely using machine learning algorithms to categorize the posts. This was mainly seen through the 'Other' topic, which often had many posts that would have been better suited with some of the other main topics. No matter how many topics we choose to have, there are always a few topics that are unpredictable and belong in the 'Other' category. For this reason, it will be near impossible to accurately predict every post into a main topic and there will certainly be a handful of posts that are misclassified as 'Other' rather than an appropriate tag.

That being said, there are a few key recommendations that we can make from the topic model as well as the Logistic Regression model.

>**Recommendations**

1) **Split up the Debt topic into multiple sub-categories**: After generating the new topics, the posts with the original 'debt' topic have split up among different topics:

In [68]:
df_eval[df_eval['old_topic']=='Debt']['new_topic'].value_counts()

loans           1077
collections      410
credit card      269
school           216
housing          165
other            143
debt             115
auto              90
employment        80
budgeting         70
insurance         70
bank account      44
planning          35
investment        23
taxes             12
Name: new_topic, dtype: int64

a) *Collections*: This was a surprisingly popular theme among quite a few posts. Out of the original 2819 posts that were dubbed as 'Debt', 410 are now part of the 'Collections'. Throughout the different machine learning models, collections consistently had among the better accuracy scores as well (0.82 precision and 0.75 recall for logistic regression model). It makes sense from a practical standpoint why collections should be added as a topic; while collections is typically a sub-category of debt, it happens frequently enough as a result of debt collection, credit freezes, scams and fraudulent activity that it should be in a category of its own. 

b) *Loans*: Even more common than collections were the posts regarding loans. Out of the original 2819 posts that were dubbed as 'Debt', 1077 are now part of the 'Loans'. These included student loans, mortgage loans, auto loans, and other common loan types. 

Overall, the goal here is to make debt more specific because of how vast the subject is and how many different types of debt there actually are on /r/personalfinance. At the same time, we don't want to make it too complex to the point that users have a difficult time categorizing their post but having just 3 types should work well. Based on the classification model performances for these sub-categories, we have also determined that they would work well for real topic assignments. 

2) **Differentiate Credit and Credit Card topics**: Credit cards are by far one of the most common subjects on /r/personalfinance. While credit cards are just one method used to build up our credit scores and improve our credit history, almost all posts under the original 'Credit' tag were about credit cards. There were almost as many posts about credit cards as there were posts categorized under the 'Credit' tag. The few posts that were not about improving their credit via using their credit cards had to deal with improving their credit through paying back their loans and bills. Issues with credit freezes would also go under the 'Collections' topic since those have to do more with the credit bureau holding a freeze on their account. In this case, it would be better to simply use a 'Credit Card' topic instead of a 'Credit' topic.

In the original set of topics, many posts regarding credit cards would sometimes come under 'Debt' despite being better suited for 'Credit'. Using a 'Credit Card' topic would help to avoid this situation. 'Debt' and 'Credit' are very similar in many different aspects so using more specific topics help to differentiate the posts.

3) **Consider introducing a new School/Student topic**: Reddit is a website dominated by people in their teens and young adult years. In fact, [42% of US internet users between the ages of 18-24](https://www.statista.com/statistics/261766/share-of-us-internet-users-who-use-reddit-by-age-group/) use Reddit and [58% of Reddit users are between the ages of 18-29](https://www.statista.com/statistics/517218/reddit-user-distribution-usa-age/). [Reddit surveys](https://imgur.com/gallery/cPzlB) also suggest that there are many Reddit users under the age of 18. Then it comes as no surprise that there is a sizeable number of posts on /r/personalfinance regarding school.

In [76]:
school_posts = len(df_eval[df_eval['new_topic']=='school'])
total_posts = len(df_eval)
school_prop = school_posts/total_posts

print('Number of School posts: {}'.format(school_posts))
print('Number of total posts: {}'.format(total_posts))
print('Proportion of School posts: {}%'.format(round(school_prop*100, 2)))

Number of School posts: 1127
Number of total posts: 23182
Proportion of School posts: 4.86%


The moderators of /r/personalfinance have even created a wiki guide-page to provide financial advice for teenagers between the ages of 15-20 and young adults between 18-25. Much of the advice deals with how to be fiscally responsible with school loans and textbooks costs. Of course, many of these questions could be regarding student loans so it may be better to categorize them under 'Loans'. 

4) **Use Machine Learning to assign topics to untagged posts using a similarity percentage threshold**: Rather than letting the system assign 'Other' to an untagged post by default, it would be better to use machine learning to assign topics.

In [56]:
other_before = len(df_eval[df_eval['old_topic']=='Other'])
other_after = len(df_eval[df_eval['new_topic']=='other'])
other_change = ((other_after - other_before)/other_before)*100
other_precision = 0.60
other_recall = 0.63

print('Number of "Other" posts with original topic set: {}'.format(other_before))
print('Number of "Other" posts with new topic set: {}'.format(other_after))
print('Change in the number of "Other" posts using the new topics: {}%'.format(round(other_change, 2)))
print('Precision Score for "Other": {}, Recall Score for "Other": {}'.format(other_precision, other_recall))

Number of "Other" posts with original topic set: 2749
Number of "Other" posts with new topic set: 2463
Change in the number of "Other" posts using the new topics: -10.4%
Precision Score for "Other": 0.6, Recall Score for "Other": 0.63


By using the newly generated topic set, there was over a 10% decrease in the number of posts tagged with 'Other' while still maintaing an overall 80% accuracy rate. To maintain as much accuracy as possible, we can adjust the model so that it will only assign a topic to an untagged post if it has over a certain percentage similarity match to a topic. For example, if we set the threshold to 50% and a post about Walgreens has a 54% match to 'Budgeting', the model will assign 'Budgeting' to be its topic. If it does not meet this threshold, the model will simply tag the post with 'Other'. This threshold can be selected by the moderators of /r/personalfinance.

>**Limitations of the Topic Model**

One pattern we did notice throughout all the models is the fact that the 'other' topic did a poor job in both precision and recall. In other words, the model sometimes falsely labeled non-'other' posts as 'other' and also did not correctly label 'other' posts as 'other'. If we had to choose between avoiding a Type I error (misclassifying 'other' posts as not 'other') and a Type II error (misclassifying non-'other' posts as 'other'), the former would be preferred. This is because the topic 'other' is a mixture of several topics whereas the other topics have a stricter definition of what kinds of posts and words belong to that category. 

For example, if an 'auto' post about car insurance was labeled as 'other', it would not be as damaging as if an 'other' post about a grandmother leaving behind a will was labeled as 'school'. There are also cases where the author of the post does not know what the topic should be; in this situation, they would most likely categorize the post under "other" rather than falsely labeling it as something definitive in nature. In a sense, the 'other' topic could be seen as a wildcard when 

Another problem does not have to do so much with the classification model, but with the topic-modeling. There are certain cases where a post was labeled under a topic that does not make sense to a human observer. Because the machine makes its decisions using vectorized inputs, it is unable to accurately classify each post 100% of the time. As a result, there are some exceptions where the classification model actually "corrects" the topic-modeling mistakes and labels some posts with the "right" topic. One explanation for this limitation is the fact that each document (or post) is assigned a number of different topics and we simply chose the topic with the highest similarity percentage as the label. This may work well for most posts but there are scenarios where this line of thinking can backfire.

For example, what if a post about college loans had a 34% contribution from 'loans' but 33% contribution from 'school'? In this case, the model would categorize this post under the 'loans' topic but the classification model may predict it as 'school'. 

>**Final Thoughts**

Topic-modeling is still very much a work-in-progress due to the fact that technology continues to adapt and evolve at a rapid pace. It especially requires an "eye-test" approach to refining and improving the models because text is not as easily interpreted by the machine compared to numbers. Therefore the user will need to verify whether or not the generated topics "work" by looking at several factors such as how much the topics make sense given the type of words that surround the topic, how many documents fit under each topic (if there are too many documents under 1 topic and too few under another, it is practically useless), etc. 

The topic model we used to generate a new set of topics for /r/personalfinance did well in categorizing posts by the main topics, such as auto and credit card. Although there were still a number of different posts that were categorized under 'Other' despite being better suited for a main topic, this model would do very well in helping the subreddit by predicting a topic for a post rather than assigning it into 'Other' by default. We also learned about possible better alternative for the current set of available topics by making them more specific.

All in all, these results looked good considering the fact that the topics were generated by the machine rather than by humans.

## 6. Resources

[Return to Table of Contents](#Table-of-Contents)

Here is a list of resources that greatly helped me learn about topic modeling and aided me in the project.

1) __[Topic Modeling and Latent Dirichlet Allocation (LDA) in Python - Susan Li](https://towardsdatascience.com/topic-modeling-and-latent-dirichlet-allocation-in-python-9bf156893c24)__

In [8]:
from IPython.display import HTML

HTML('''<script>
code_show=true; 
function code_toggle() {
 if (code_show){
 $('div.input').hide();
 } else {
 $('div.input').show();
 }
 code_show = !code_show
} 
$( document ).ready(code_toggle);
</script>
<form action="javascript:code_toggle()"><input type="submit" value="Click here to toggle on/off the raw code."></form>''')
